### PR TITLE
Make disableCircularProxies() disable all forms of circular dependencies.

### PR DESCRIPTION
**Pull request by tavianator**
This addresses https://github.com/google/guice/issues/970.  It's an incomplete fix: there's no pre-validation of the graph to check for cycles -- all the checking happens during previsioning.  But I wanted to post it early to get feedback on the approach (and to get Travis to run it).  I'll add the early validation in a later ~~commit~~ pull request.

### CODE REVIEW

File: core/src/com/google/inject/Binder.java [hunk 1]
```diff
@@ -487,6 +487,16 @@ void bindListener(Matcher<? super TypeLiteral<?>> typeMatcher,
...
   
   /**
+   * Prevents Guice from injecting dependencies that form a cycle, unless broken by a
+   * {@link Provider}. By default, circular dependencies are not disabled.
+   * <p>
+   * This implies {@link #disableCircularProxies()}.
+   *
+   * @since 5.0
+   */
+  void disableCircularDependencies();
```
Comment (sameb) [inline core/src/com/google/inject/Binder.java]: FWIW, I'd just kill this and have disableCircularProxies imply it.

Comment (tavianator) [inline core/src/com/google/inject/Binder.java]: Done!

File: core/src/com/google/inject/internal/Errors.java [hunk 2]
```diff
@@ -450,15 +450,15 @@ public Errors cannotInjectRawTypeLiteral() {
...
     return addMessage(
         "Tried proxying %s to support a circular dependency, but it is not an interface.",
         expectedType);
   }
 
-  public Errors circularProxiesDisabled(Class<?> expectedType) {
+  public Errors circularDependenciesDisabled(Class<?> expectedType) {
     return addMessage(
-        "Tried proxying %s to support a circular dependency, but circular proxies are disabled.",
+        "Found a circular depenency involving %s, and circular dependencies are disabled.",
```
Comment (sameb) [inline core/src/com/google/inject/internal/Errors.java]: s/depenency/dependency

**Comment by sameb** [comment]
This LGTM -- I'd prefer actually committing this as-is and doing the eager checking in another PR as a followup.  That sound good to you?

File: core/src/com/google/inject/internal/Errors.java (hunk 2 above)
Comment (tavianator) [inline core/src/com/google/inject/internal/Errors.java]: Oops, fixed :)

**Comment by tavianator** [comment]
@sameb Yeah that sounds good.

**Comment by sameb** [comment]
Will merge once Travis comes back clean, thanks @tavianator!

**Comment by tavianator** [comment]
@sameb Looks like Travis came back clean :)

**Comment by sameb** [comment]
Thanks for the reminder!

**Pull request merged**